flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java

```diff
@@ -15,75 +16,474 @@
+			new HashMap<String, Set<Entry<GlobalStreamId, Grouping>>>();
+
+	final HashMap<String, HashMap<String, DataStream<Tuple>>> availableInputs = new HashMap<>();
 
-	/** The number of declared tasks for the whole program (ie, sum over all dops) */
-	private int numberOfTasks = 0;
+	private final TopologyBuilder builder;
 
-	public FlinkTopology() {
-		// Set default parallelism to 1, to mirror Storm default behavior
-		super.setParallelism(1);
+	// needs to be a class member for internal testing purpose
+	private final StormTopology stormTopology;
+
+	private final Map<String, IRichSpout> spouts;
+	private final Map<String, IRichBolt> bolts;
+
+	private final StreamExecutionEnvironment env;
+
+	private FlinkTopology(TopologyBuilder builder) {
+		this.builder = builder;
+		this.stormTopology = builder.createTopology();
+		// extract the spouts and bolts
+		this.spouts = getPrivateField("_spouts");
+		this.bolts = getPrivateField("_bolts");
+
+		this.env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		// Kick off the translation immediately
+		translateTopology();
 	}
 
 	/**
-	 * Is not supported. In order to execute use {@link FlinkLocalCluster}, {@link FlinkSubmitter}, or {@link
-	 * FlinkClient}.
 	 *
-	 * @throws UnsupportedOperationException
-	 * 		at every invocation
+	 * Creates a Flink program that uses the specified spouts and bolts.
+	 * @param stormBuilder The storm topology builder to use for creating the Flink topology.
+	 * @return A Flink Topology which may be executed.
 	 */
-	@Override
-	public JobExecutionResult execute() throws Exception {
-		throw new UnsupportedOperationException(
-				"A FlinkTopology cannot be executed directly. Use FlinkLocalCluster, FlinkSubmitter, or FlinkClient " +
-				"instead.");
+	public static FlinkTopology createTopology(TopologyBuilder stormBuilder) {
+		return new FlinkTopology(stormBuilder);
 	}
 
 	/**
-	 * Is not supported. In order to execute use {@link FlinkLocalCluster}, {@link FlinkSubmitter} or {@link
-	 * FlinkClient}.
-	 *
-	 * @throws UnsupportedOperationException
-	 * 		at every invocation
+	 * Returns the underlying Flink ExecutionEnvironment for the Storm topology.
+	 * @return The contextual environment.
 	 */
-	@Override
-	public JobExecutionResult execute(final String jobName) throws Exception {
-		throw new UnsupportedOperationException(
-				"A FlinkTopology cannot be executed directly. Use FlinkLocalCluster, FlinkSubmitter, or FlinkClient " +
-				"instead.");
+	public StreamExecutionEnvironment getExecutionEnvironment() {
+		return this.env;
 	}
 
 	/**
-	 * Increased the number of declared tasks of this program by the given value.
-	 *
-	 * @param dop
-	 * 		The dop of a new operator that increases the number of overall tasks.
+	 * Directly executes the Storm topology based on the current context (local when in IDE and
+	 * remote when executed thorugh ./bin/flink).
+	 * @return The execution result
+	 * @throws Exception
 	 */
```

Comment:
Can we describe the exception in a meaningful way? Or are there too many reasons to get listed here? What does JavaDoc of `env.execute()` say?